### PR TITLE
feat: projection adapter + shadow harness

### PR DESCRIPTION
**Pull request by matthiasn**
## Summary by CodeRabbit

* **New Features**
  * Read-only shadow comparison reporting match, forked, mismatch, empty, or error with diagnostic details and projected head IDs.
* **Chores**
  * Convert persisted agent message logs into kernel events with sensible defaults for host, vector clock, kind, and causal parents.
* **Bug Fixes**
  * More robust handling of dangling/invalid parent links and cycle-related errors during projection.
* **Tests**
  * Expanded unit and property tests for event mapping, causal-parent reconstruction, and shadow comparison.
* **Documentation**
  * Updated roadmap and projection plan to adopt per-host G-counter design and sequencing.

[Review Change Stack](https://app.coderabbit.ai/change-stack/matthiasn/lotti/pull/3246?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/features/agents/projection/shadow_projection.dart [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    );
+  } on Exception catch (e) {
+    return ShadowProjectionReport(
+      status: ShadowProjectionStatus.error,
+      projectedHeadIds: const [],
+      liveHeadId: liveHeadId,
+      danglingParentIds: const [],
+      error: e.toString(),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/agents/projection/shadow_projection.dart]: Fixed — changed to a bare `catch (e)` so `Error`s (StateError/TypeError/AssertionError) are captured too, not just `Exception`. A structural failure in canonicalOrder/project now always surfaces as `status: error` instead of propagating, honoring the never-crash-a-production-path goal. (Added a comment explaining the broad catch; the repo disables `avoid_catches_without_on_clauses`, so no ignore is needed.)

File: lib/features/agents/projection/shadow_projection.dart [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  if (heads.isEmpty) {
+    return liveHeadId == null
+        ? ShadowProjectionStatus.empty
+        : ShadowProjectionStatus.mismatch;
+  }
+  if (heads.length > 1) return ShadowProjectionStatus.forked;
+  return heads.single == liveHeadId
+      ? ShadowProjectionStatus.match
+      : ShadowProjectionStatus.mismatch;
+}
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/agents/projection/shadow_projection.dart:125]: Fixed. Added `if (liveHeadId == null) return mismatch;` after the empty check, so a non-empty projection with no live head pointer is a mismatch regardless of head count — `forked` now only means 'live tracks one of several tips'. Tightened the enum docs and added a deterministic test (fork + null live head → mismatch, not forked) covering the new branch.

File: lib/features/agents/projection/shadow_projection.dart [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+/// reported as [ShadowProjectionStatus.error].
+ShadowProjectionReport compareShadowProjection({
+  required Iterable<AgentMessageEntity> messages,
+  required Iterable<AgentLink> links,
+  required String? liveHeadId,
+  String Function(AgentMessageEntity message)? hostIdOf,
+}) {
+  final events = agentEventsFromLog(messages, links, hostIdOf: hostIdOf);
+  try {
+    final projection = project(canonicalOrder(events));
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/agents/projection/shadow_projection.dart]: Fixed — moved `agentEventsFromLog(...)` inside the `try`. `messages`/`links` may be lazy/DB-backed iterables and `hostIdOf` is caller-supplied, so building the events can now also surface as `status: error` rather than propagating, fully honoring never-crash-a-production-path.

File: lib/features/agents/projection/shadow_projection.dart [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+ShadowProjectionStatus _statusFor(List<String> heads, String? liveHeadId) {
+  if (heads.isEmpty) {
+    return liveHeadId == null
+        ? ShadowProjectionStatus.empty
+        : ShadowProjectionStatus.mismatch;
+  }
+  // A non-empty projection with no live head pointer is a genuine mismatch —
+  // not the `forked` divergence, which is "live tracks one of several tips".
+  if (liveHeadId == null) return ShadowProjectionStatus.mismatch;
+  if (heads.length > 1) return ShadowProjectionStatus.forked;
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/agents/projection/shadow_projection.dart]: Fixed. `forked` now requires the live head to actually be one of the projected tips. Restructured `_statusFor` to: non-empty heads + (liveHeadId null OR not in heads) → `mismatch`; otherwise `length > 1 ? forked : match`. So a live head pointing at a stale parent (non-tip) is a mismatch, not a fork.

File: test/features/agents/projection/shadow_projection_test.dart [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+          makeTestMessage(id: 'b'),
+          makeTestMessage(id: 'c'),
+        ],
+        [_prevLink('b', 'a'), _prevLink('c', 'a')],
+        'b', // live tracks one of the two heads
+      );
+
+      expect(report.status, ShadowProjectionStatus.forked);
+      expect(report.projectedHeadIds.toSet(), {'b', 'c'});
+    });
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/agents/projection/shadow_projection_test.dart:88]: Added the example test (live tracks parent 'a', tips are {b,c} → mismatch). Also strengthened the property test: it now varies `liveHeadId` across tip / non-tip / absent / null and asserts the status for each category over generated DAGs, so the non-tip mismatch path is covered generatively, not just by the example.
